src/character_management.cpp

```diff
@@ -0,0 +1,155 @@
+#include "character_management.h"
+
+#include "character.h"
+#include "item.h"
+#include "output.h"
+
+static bool has_time_passed( const item &it, const std::string &var, const time_duration &interval )
+{
+    return calendar::turn - it.get_var( var, calendar::turn - 2 * interval ) > ( interval - 1_seconds );
+}
+
+static std::optional<int> reload_options_menu( const std::vector<item_location> &mags,
+        const std::string it_name )
+{
+    int num_mags = mags.size();
+
+    uilist rmenu;
+    rmenu.text = _( string_format( "Reload your %s with what?", it_name ) );
+    // To avoid having duplicate entries in the menu, only add the first entry with a unique name
+    std::set<std::string> added_names;
+    for( int i = 0; i < num_mags; ++i ) {
+        std::string name = mags[i]->display_name();
+        // Duplicate entry
+        if( added_names.count( name ) != 0 ) {
+            continue;
+        }
+        added_names.emplace( name );
+        rmenu.addentry( i, true, MENU_AUTOASSIGN, name );
+    }
+    // Give them an option to chose nothing, too
+    rmenu.addentry( num_mags, true, MENU_AUTOASSIGN, _( "Don't reload" ) );
+    // Now, see what they want
+    rmenu.query();
+    // If they picked one of the valid mags
+    if( rmenu.ret >= 0 && rmenu.ret < num_mags ) {
+        return std::optional<int>( rmenu.ret );
+    }
+    return std::nullopt;
+}
+
+bool Character::try_autoreload( item &it, const char_autoreload::params &info )
+{
+    const std::string fail_var_name = "autoreload_failed_time";
+    const std::string warn_var_name = "autoreload_warn_time";
+    const bool is_npc = !is_avatar();
+
+    // Just get the name once
+    std::string it_name = it.tname();
+
+    // If it's not too low, we'll just let it be
+    int uses_left = it.ammo_remaining();
+    if( uses_left > 9 ) {
+        return false;
+    }
+
+    // And we'll only warn unless we're just about to run out
+    if( uses_left > 1 ) {
+        if( !is_npc && has_time_passed( it, warn_var_name, 8_seconds ) ) {
+            add_msg_if_player( m_info, _( "Your %s is running low" ), it_name );
+            it.set_var( warn_var_name, calendar::turn );
+        }
+        return false;
+    }
+
+    // If we tried to reload it recently but couldn't, avoid spamming the player
+    if( it.has_var( fail_var_name ) && !has_time_passed( it, fail_var_name, 1_minutes ) ) {
+        return false;
+    }
+
+    // Now that we're here, we've decided to reload
+    // So, find all acceptable reloadables in the character inventory
+    std::vector<item_location> mags;
+    for( item_location &loc : all_items_loc() ) {
+        if( !it.can_reload_with( *loc, false ) ) {
+            continue;
+        }
+        // Only mags with sufficient ammo to avoid immediate reload, too!
+        if( loc->ammo_remaining() < 2 ) {
```

Comment:
Might also make sense to auto-reload if the current magazine is depleted but there are magazines with less than this many remaining charges.